cty/value_ops.go

```diff
@@ -74,9 +74,13 @@ func (val Value) GoString() string {
 // as dynamic if either of the given values are. Use RawEquals to compare
 // if two values are equal *ignoring* the short-circuit rules.
 func (val Value) Equals(other Value) Value {
+	// Nulls are equal, regardless of type.
+	if val.IsNull() && other.IsNull() {
+		return BoolVal(true)
+	}
+
 	// If only one is null, then we know this must be false
-	// val.IsNull() XOR other.IsNull()
-	if (val.IsNull() || other.IsNull()) && !(val.IsNull() && other.IsNull()) {
+	if val.IsNull() || other.IsNull() {
```

Comment:
Since this check is relying on the branch above, I think it could be easier to understand the control flow here to have these grouped together as a single construct that deals with everything null-related:

```go
switch {
case val.IsNull() && other.IsNull():
    // Nulls are always equal, regardless of type
    return BoolVal(true)
case val.IsNull() || other.IsNull():
    // If only one is null then the result must be false
    return BoolVal(false)
}
```

(or an `if`, `else if` chain, if you prefer.)